docs/RFCS/time_series_culling.md

```diff
@@ -0,0 +1,349 @@
+in order to avoid duplicate work; however, this is *not* necessary for
+correctness, as demonstrated in the [Rollup Algorithm](#rollup-algorithm)
+section below.
+
+The queue will initially be set to process replicas at the same rate as the
+replica GC queue (as of this RFC, one range per 50 milliseconds).
+
+##### Package Dependency
+
+There is one particular complication to this method: *go package dependency*.
+Knowledge on how to identify and cull time series keys is contained in the `ts`
+package, but all logic for replica queues (and all current queues) lives in
+`storage`, meaning that one of three things must happen:
+
++ `storage` can depend on `ts`. This seems to be trivially possible now, but may
+be unintuitive to those trying to understand our code-base. For reference, the
+`storage` package used to depend on the `sql` package in order to record event
+logs, but this eventually became an impediment to new development and had to be
+modified.
++ The queue logic could be implemented in `ts`, and `storage` could implement
+an interface that allows it to use the `ts` code without a dependency.
++ Parts of the `ts` package could be split off into another package that can
+intuitively live below `storage`. However, this is likely to be a considerable
+portion of `ts` in order to properly implement rollups.
+
+Tenatively, we will be attempting to use the first method and have `storage`
+depend on `ts`; if it is indeed trivially possible, this will be the fastest
+method of completing this project.
+
+#### Culling low resolution data
+
+Although the volume is much lower, low-resolution data will still build
+up indefinitely unless it is culled. This data will also be culled by the same
+algorithm outlined here; however, it will not be rolled up further, but will
+simply be deleted.
+
+## Rollup algorithm
+
+The rollup algorithm is intended to be run on a single high-resolution key
+identified by the culling algorithm. The algorithm is as follows:
+
+1. Read the data in the key. Each key represents a "slab" of high resolution
+samples captured over a wall-clock hour (up to 360 samples per hour).
+2. "Downsample" all of the data in the key into a single sample; the new sample
+will have a sum, count, min and max, computed from the samples in the original
+key.
+3. Write the computed sample as a low-resolution data point into the time series
+system; this is exactly the same process as currently recorded time series,
+except it will be writing to a different key space (with a different key
+prefix).
+4. Delete the original high-resolution key.
+
+This algorithm is safe to use, even in the case where the same key is being
+culled by multiple nodes at the same time; this is because step 3 and 4 are
+currently *idempotent*. The low-resolution sample generated by each node will be
+identical, and the engine-level time series merging system currently discards
+duplicate samples. The deletion of the high-resolution key may cause an error on
```

Comment:
Ah, this seems to deal with some of it. Are we requiring that the top-boundary of the key is there? Or are we simply requiring that the result set is not empty?